teps/0052-tekton-results-automated-run-resource-cleanup.md

```diff
@@ -0,0 +1,213 @@
+
+Many users have implemented custom cleanup mechanisms to address this
+([including our own dogfooding cluster](https://github.com/tektoncd/plumbing/blob/2c1808d75b38d6fcaae0a37f5a9dda50545af4c8/tekton/resources/cd/cleanup-template.yaml)).
+This proposal aims to define a supported mechanism for this behavior in
+conjunction with our existing Results storage.
+
+Relevant Issues:
+
+- [tektoncd/pipeline#2332](https://github.com/tektoncd/pipeline/issues/2332)
+- [tektoncd/pipeline#1486](https://github.com/tektoncd/pipeline/pull/1486)
+- [tektoncd/pipeline#1429](https://github.com/tektoncd/pipeline/pull/1429)
+- [tektoncd/experimental#479](https://github.com/tektoncd/experimental/issues/479)
+- [tektoncd/pipeline#2452](https://github.com/tektoncd/pipeline/issues/2452)
+
+### Goals
+
+- Define a mechanism for cleaning up completed Runs to minimize resources the
+  Pipeline controller is responsible for.
+- Preserve Run metadata so that users can continue to query it beyond the
+  lifetime of the Pipeline controller.
+
+### Non-Goals
+
+- Providing a programmatic mechanism for end users to dynamically control deletion behavior.
+
+### Use Cases (optional)
+
+This is a commonly requested feature due to the strain Run history this puts on
+etcd over time. See [Relevant Issues under Motivation](#motivation).
+
+## Requirements
+
+- Run metadata should be available even after deletion from the Pipeline
+  controller data plane.
+- Runs should be deleted as soon as it can be done safely (i.e. we have
+  confirmed the data is stored durably elsewhere).
+- Deletion should occur with some configurable delay to allow other integrations
+  watching Runs to respond to events.
+- Operators should be able to opt in / out to deletion behavior.
+
+## Proposal
+
+### Delete Runs when complete
+
+The
+[Tekton Results watcher](https://github.com/tektoncd/results/blob/main/docs/watcher/README.md)
+already watches for changes to TaskRun and PipelineRuns to upload Records for
+long term storage. Once we have verified that 1)
+[the Run has completed](https://github.com/tektoncd/pipeline/blob/master/docs/runs.md#monitoring-execution-status)
+and 2) the Run Record has been successfully stored, we know we can delete the
+Run without affecting execution.
+
+### Grace Period
+
+Tekton Results may not be the only controller listening to changes in TaskRun
+and PipelineRun CRDs (e.g. Notifications that happen post-Run execution).
+Because of this, deleting the Run immediately might cause unintended race
+conditions and affect other integrations using a similar mechanism.
+
+To address this, we should have a user configurable grace period for Run
+deletion, and require that the time elapsed since the Record's last update time
+(since the k8s ObjectMeta does not have a similar concept of last update time)
+is greater than the configured grace period before deletion. If the grace period
+has not been met yet, then we should simply
```

Comment:
can you explain a bit more about what you mean by "the grace period has not been met yet"?